### PR TITLE
Improve syntaxes

### PR DESCRIPTION
**Pull request by bynect**
Improve existing syntaxes, add Rust syntax descriptor, add number suffix highlighting and fix number prefix highlighting

### CODE REVIEW

**Comment by arthurbacci** [comment]
Thanks for the Pull Request, but it will not be accepted now because it is extremely slow with large files.

**Comment by bynect** [comment]
> Thanks for the Pull Request, but it will not be accepted now because it is extremely slow with large files.

What part made it slower precisely?

**Comment by arthurbacci** [comment]
> > Thanks for the Pull Request, but it will not be accepted now because it is extremely slow with large files.
> 
> What part made it slower precisely?

I don't know. I just tested it and it is very laggy in the end of the file.

**Comment by arthurbacci** [comment]
I think it can be faster if instead of calling `syntaxHighlight` for every move, calling it only when the scroll changes. I will make a commit now.

**Comment by bynect** [comment]
> > > Thanks for the Pull Request, but it will not be accepted now because it is extremely slow with large files.
> > 
> > 
> > What part made it slower precisely?
> 
> I don't know. I just tested it and it is very laggy in the end of the file.

This problem was already there, I tried a file with ~4k lines and jumping to the end is all super laggy

**Comment by bynect** [comment]
> I think it can be faster if instead of calling `syntaxHighlight` for every move, calling it only when the scroll changes. I will make a commit now.

If I remember correctly I made that change some days ago, anyways the problem here is that every syntaxHighlight call is looping over all the document for strings/multiline comments

**Comment by bynect** [comment]
The only way to fix that slowness is to remove that loop and create some mechanism that stores comment/string information on the line itself IMHO

**Comment by arthurbacci** [comment]
I think we can calculate strings and that sort of things only if that line is visible, and invisible lines should only be used for detecting multiline comments.

**Comment by bynect** [comment]
> I think we can calculate strings and that sort of things only if that line is visible, and invisible lines should only be used for detecting multiline comments.

The fact is that with multiline comments and strings we can't know if a line not shown starts one of them without either looping (aka checking every time, slow) or storing the information somewhere else

**Comment by arthurbacci** [comment]
the0problem
0 is white.

I will try fixing it now.

**Comment by bynect** [comment]
> the0problem
> 0 is white.
> 
> I will try fixing it now.

Fixed, I erroneously removed an if check

**Comment by arthurbacci** [comment]
What about fixing the lag here before merging?

**Comment by arthurbacci** [comment]
I think a system like this would be good:

```
0 "I don't have a comment"
1 "I have a comment /*"
1 "I am commented"
1 "I am commented"
0 "Comment closes here */"
```

With it, we can highlight comments without calculating the syntax for all lines.

**Comment by arthurbacci** [comment]
Before implementing it I will try to remove some things that don't need to be done with the invisible lines.

**Comment by bynect** [comment]
> I think a system like this would be good:
> 
> ```
> 0 "I don't have a comment"
> 1 "I have a comment /*"
> 1 "I am commented"
> 1 "I am commented"
> 0 "Comment closes here */"
> ```
> 
> With it, we can highlight comments without calculating the syntax for all lines.

I also taught about that, a sort of flag for every line that tells either if the line is a comment or a string. This way the only time when we need to recalculate the syntax and update those flags is when we modify the file

**Comment by arthurbacci** [comment]
Did some changes, speed is far better now. I will change some other things and commit.

**Comment by bynect** [comment]
tednum
Seems to finally be working properly, also idk why git forcepushed

File: src/color.c [hunk 1]
```diff
@@ -20,10 +20,7 @@ void syntaxHighlight(void) {
     unsigned int octprefixlen = strlen(config.current_syntax->number_prefix[1]);
     unsigned int binprefixlen = strlen(config.current_syntax->number_prefix[2]);
 
-    unsigned int syntax_start = 0;
-    unsigned int syntax_end = num_lines;
-
-    for (unsigned int at = syntax_start; (at < syntax_end) && (at != num_lines); ++at) {
+    for (unsigned int at = 0; num_lines; at++) {
```
Comment (bynect) [inline src/color.c]: Isn't the condition missing here?

Comment (arthurbacci) [reply src/color.c]: Yes, I fixed it.

**Comment by bynect** [comment]
Just tried to jump to the end of a file with ~35k lines and ted is not laggy like before

**Comment by bynect** [comment]
When writing to or opening large files this new approach is very expensive and on that large file writing a single letter was very laggy and opening it required several seconds, so we should optimize even more to not waste time and resources on lines that are not visible

**Pull request merged**